beamer/agent/chain.py

```diff
@@ -572,6 +575,61 @@ def _timestamp_is_l1_finalized(
     return int(time.time()) > timestamp + target_chain_finality
 
 
+def maybe_prove(claim: Claim, context: Context) -> bool:
+    request = context.requests.get(claim.request_id)
+
+    assert request is not None, "Active claim for non-existent request"
+
+    # mainnet: 10, goerli: 420, local: 17
+    if context.target_chain_id not in [10, 420, 17]:
+        return False
+
+    own_challenge_stake = Wei(claim.get_challenger_stake(context.address))
+    can_prove = (
+        claim.claimer == context.address and claim.challenger_exists()
+    ) or own_challenge_stake > 0
+
+    if not can_prove:
+        return False
+
+    if claim.message_proved:
+        return False
+
+    if request.fill_tx:
+        proof_tx = request.fill_tx
+    elif claim.invalidation_tx:
+        proof_tx = claim.invalidation_tx
+
+    if proof_tx in context.l1_resolutions:
+        return False
+
+    future = context.task_pool.submit(
+        run_relayer_for_tx,
+        context.config.base_chain_rpc_url,
+        context.target_rpc_url,
+        context.source_rpc_url,
+        context.config.account.key,
+        proof_tx,
+        True,
+    )
+
+    def on_future_done(f: Future) -> None:
+        try:
+            f.result()
+            claim.message_proved = True
+        except Exception as ex:
+            context.logger.error("Optimism prove failed", ex=ex, tx_hash=proof_tx)
+        finally:
+            context.l1_resolutions.pop(proof_tx, None)
+
+    future.add_done_callback(on_future_done)
+    context.l1_resolutions[proof_tx] = future
+
+    context.logger.info("Initiated Optimism prove", request=request, claim=claim, tx_hash=proof_tx)
```

Comment:
Since this is a call on L1 and the fact that it's not easy to debug if things go wrong here, should we perhaps consider adding a check for roughly enough balance on L1, before we invoke the relayer?

Reply:
I remember we would increase the claimStake and changing the finality period for optimism and whenever there is a challenge agent will send a prove to actually start 7 day period, if we wait for balance here then it wouldnt be possible right @fredo  ?

Reply:
claimStake and the actual balance on L1 is something different.

The claimStake just makes sure that the agent will get reimbursed. For the prove transaction we decided to not do any checks regarding financial viability.

Regarding balance on L1, I don't think the way we separated tasks between the relayer and the agent, I would actually prefer to not do the check here in the agent. IMO it's the relayer's job to communicate correctly what happened when trying to execute the task.
I'd rather say that if we want to have better logging of whats going on in the relayer, we should think about communicating more via exit codes or exceptions.

Reply:
My concern here is wiith this:

> agent will send a prove

The agent can only successfully  do that if there are enough funds on L1. Otherwise, the relayer will fail and we will have no way of knowing why, right? It would be nice if we captured at least `stderr` of the relayer process and dumped it into the agent's log.

Reply:
If we go with that we should spin out its own issue for it, and provide a detailed design for communication between relayer and agent via exit codes or exceptions

Reply:
>  I'd rather say that if we want to have better logging of whats going on in the relayer, we should think about communicating more via exit codes or exceptions.

:this: